### PR TITLE
Afbase/aturi validation rsky syntax

### LINKED ISSUE #38: Proposal: Fully featured rsky-syntax
Working Branch: [afbase/aturi-validation-rsky-syntax](https://github.com/afbase/rsky/tree/afbase/aturi-validation-rsky-syntax)

# Complete Implementation of AT Protocol String Types in rsky-syntax

## Overview
This PR completes the Rust implementation of AT Protocol string type validation, matching the reference TypeScript implementation. It adds support for DID, Handle, NSID, TID, Record Key, and DateTime validation while making all string functions generic over `Into<String>`.

All functions maintain backward compatibility while adding support for generic string types.

### Added Files
- `rsky-syntax/src/aturi_validation.rs` - AT URI validation implementation
- `rsky-syntax/src/datetime.rs` - DateTime validation implementation
- `rsky-syntax/src/did.rs` - DID validation implementation
- `rsky-syntax/src/nsid.rs` - NSID validation implementation 
- `rsky-syntax/src/record_key.rs` - Record Key validation implementation
- `rsky-syntax/src/tid.rs` - TID validation implementation

### Modified Files
- `rsky-syntax/Cargo.toml` - Added chrono dependency
- `rsky-syntax/src/lib.rs` - Added new module exports
- `rsky-syntax/src/aturi.rs` - Generic string handling + string conversion traits
- `rsky-syntax/src/handle.rs` - Generic string handling
- `rsky-pds/src/repo/record/mod.rs` - Added validation checks

### 1. Generic String Handling
All validation functions now accept any type that implements `Into<String>`, making them more flexible while maintaining zero-cost abstractions. For example:

```rust
// Using ensure_valid_did with different string types
use rsky_syntax::did::ensure_valid_did;

// With string literal
ensure_valid_did("did:plc:7iza6de2dwap2sbkpav7c6c6")?;

// With owned String 
let did = String::from("did:plc:7iza6de2dwap2sbkpav7c6c6");
ensure_valid_did(did)?;

// From an AtUri
let uri = AtUri::new("at://did:plc:7iza6de2dwap2sbkpav7c6c6".to_string(), None)?;
ensure_valid_did(uri.host)?;
```

### 2. Complete Type Coverage
Implemented all AT Protocol string types:
- DID validation (did.rs)
- Handle validation (handle.rs) 
- NSID validation (nsid.rs)
- TID validation (tid.rs)
- Record Key validation (record_key.rs)
- DateTime validation (datetime.rs)
- AT URI validation (aturi.rs, aturi_validation.rs)

### AT URI Validation 
The TypeScript codebase currently has an interesting pattern - it exports AT URI validation functions (`ensureValidAtUri`, `ensureValidAtUriRegex`) but doesn't use them in the AT URI parser. This leads to two questions:

1. Should we validate AT URIs in our codebase even though TypeScript doesn't? The validation code exists but isn't used.

2. Our `ensure_valid_at_uri` returns `Result<AtUri, AtUriValidationError>` rather than just `Result<(), AtUriValidationError>`. This is a slight enhancement over TypeScript as it allows validation and parsing in one step:

```rust
// TypeScript-style (two steps)
ensure_valid_at_uri(&uri)?;
let at_uri = AtUri::try_from(uri)?;

// New Rust-style (one step)
let at_uri = ensure_valid_at_uri(uri)?;
```

## Testing
Comprehensive test suite added, matching TypeScript test cases. All validation functions have corresponding tests verifying:
- Valid input acceptance
- Invalid input rejection
- Edge cases
- Real-world examples

## Next Steps
1. Decision needed on AT URI validation strategy
2. Consider whether to align `ensure_valid_at_uri` return type with TypeScript or keep the enhanced version
3. PR 🎊

### PR DESCRIPTION
**Pull request by afbase**
Closes #38 

>Should we validate AT URIs in our codebase even though TypeScript doesn't? The validation code exists but isn't used.

Okay upon further testing and review, the [ensureValidAtUri](https://github.com/bluesky-social/atproto/blob/main/packages/syntax/src/aturi_validation.ts#L18)  in the typscript codebase does not consider:
1. AtURIs with the query parameters.  
2. As far as I can tell, it is _only_ used to validate the "subject" URIs that you may find in repository records.
3. My implementation conforms to the typescript capability and so I have decided to not incorporate it in `AtUri::new` code.

You'll still be able to:
```rust
// New Rust-style (one step)
let at_uri = ensure_valid_at_uri(uri)?;
// or...
let _ = ensure_valid_at_uri(uri)?;
```

and similarly in the other validation functions implementations, `ensure_valid_datetime`, if ok, will return a `DateTime<FixedOffset>>`.  So you'll be able to:
```rust
// New Rust-style (one step)
let validated_datetime = ensure_valid_datetime(date_str)?;
// or...
let _ = ensure_valid_datetime(date_str)?;
```

I'm doing so since the code parses the string to a datetime object successfully.  The other validation code evaluates strings and it's a bit redundant to return a string again.

### CODE REVIEW

File: rsky-syntax/src/aturi_validation.rs [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+        expect_invalid("at://did:plc:asdf123/com.atproto.feed.post/record/#/frag");
+    }
+
+    #[test]
+    fn test_record_keys() {
+        // Test valid record keys
+        expect_valid("at://did:plc:asdf123/com.atproto.feed.post/asdf123");
+        expect_valid("at://did:plc:asdf123/com.atproto.feed.post/a");
+        expect_valid("at://did:plc:asdf123/com.atproto.feed.post/%23");
+        expect_valid("at://did:plc:asdf123/com.atproto.feed.post/$@!*)(:,;~.sdf123");
```
Comment (rudyfraser) [inline rsky-syntax/src/aturi_validation.rs:291]: Some of these don't look like valid record keys to me. Per [the docs](https://atproto.com/specs/record-key):

> Regardless of the type, Record Keys must fulfill some baseline syntax constraints:
> 
>     - restricted to a subset of ASCII characters — the allowed characters are alphanumeric (A-Za-z0-9), period, dash, underscore, colon, or tilde (.-_:~)

dollar signs, at signs, parentheses, semi-colons, asterisks, and exclamation points shouldn't be allowed.

File: rsky-syntax/src/aturi_validation.rs [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+        expect_valid("at://did:plc:asdf123/com.atproto.feed.post/%%%");
+    }
+
+    #[test]
+    fn test_length_limits() {
+        // Test valid length
+        expect_valid("at://did:plc:asdf123/com.atproto.feed.post/record");
+        expect_valid(&format!(
+            "at://did:plc:asdf123/com.atproto.feed.post/{}",
+            "o".repeat(800)
```
Comment (rudyfraser) [inline rsky-syntax/src/aturi_validation.rs:340]: must have at least 1 and at most 512 characters per [the docs](https://atproto.com/specs/record-key)

File: rsky-syntax/src/aturi_validation.rs (hunk 1 above)
Comment (afbase) [reply rsky-syntax/src/aturi_validation.rs:357]: would  you like me conform more closely to the docs or to what typescript has?

Typescript's validation is much more relaxed at this time 😞

**Comment by afbase** [comment]
@rudyfraser - give me a another day or so to work on this.  There are some discrepancies between the typescript and atproto specifications.

1. I'm going to make a markdown file of the discrepancies in the validation between this code, the typescript, and the atproto specs.
2. I reworked the aturi validation but I'm going to go through the implementations and tests of the other files again to make sure I'm content with it.

**Comment by rudyfraser** [comment]
> @rudyfraser - give me a another day or so to work on this. There are some discrepancies between the typescript and atproto specifications.
> 
>     1. I'm going to make a markdown file of the discrepancies in the validation between this code, the typescript, and the atproto specs.
> 
>     2. I reworked the aturi validation but I'm going to go through the implementations and tests of the other files again to make sure I'm content with it.
> 
> 
> I'll let you know when it's ready for review again

Sounds good. Just to answer your other comment the atproto specs should be the source of truth where there are differences between the spec and the TS code. I imagine discrepancies should be rare (since it's the same org working on both, lol) but for a case like this it'd be the spec.

I'll keep an eye out for your next comment to review.

**Comment by afbase** [comment]
> > @rudyfraser - give me a another day or so to work on this. There are some discrepancies between the typescript and atproto specifications.
> > ```
> > 1. I'm going to make a markdown file of the discrepancies in the validation between this code, the typescript, and the atproto specs.
> > 
> > 2. I reworked the aturi validation but I'm going to go through the implementations and tests of the other files again to make sure I'm content with it.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I'll let you know when it's ready for review again
> 
> Sounds good. Just to answer your other comment the atproto specs should be the source of truth where there are differences between the spec and the TS code. I imagine discrepancies should be rare (since it's the same org working on both, lol) but for a case like this it'd be the spec.
> 
> I'll keep an eye out for your next comment to review.

Okay so here's what I think is going on reviewing the typescript code and specification more thoroughly:

1. It seems there are two AT-URI syntaxes, a general form and the more restricted form that is meant for the use with lexicon.  This validation seems to follow the general form (and the typescript implementation as well) - which is a quite a bit more generous with what is acceptable.
2. With that said, the AT-URI validation here is fairly weak.  See this [note](https://github.com/bluesky-social/atproto/blob/main/packages/syntax/tests/aturi.test.ts#L524) in the typescript implementation:
>NOTE: this package is currently more permissive than spec about AT URIs, so invalid cases are not errors

3. The other components, i.e. DID, Handle, datetime, NSID, TID, and Record Key, seem to conform to the specifications on atproto.com.   
4. I made `rsky-syntax/ATURI_VALIDATION.md` which provides some details about this for others to get some idea about the details about this.

**Comment by afbase** [comment]
An idea: We could create a validation function that is more strict and conform to the specification. This would got beyond the reference implementation in Typescript.

**Comment by rudyfraser** [comment]
For now could you open an issue for the more strict version? Since this conforms to the canonical one I think we're good to go here.
